File: common/src/main/java/io/github/fisher2911/hmcleaves/HMCLeaves.java
/*
 *
 *  *     HMCLeaves
 *  *     Copyright (C) 2022  Hibiscus Creative Studios
 *  *
 *  *     This program is free software: you can redistribute it and/or modify
 *  *     it under the terms of the GNU General Public License as published by
 *  *     the Free Software Foundation, either version 3 of the License, or
 *  *     (at your option) any later version.
 *  *
 *  *     This program is distributed in the hope that it will be useful,
 *  *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  *     GNU General Public License for more details.
 *  *
 *  *     You should have received a copy of the GNU General Public License
 *  *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

package io.github.fisher2911.hmcleaves;

import com.github.retrooper.packetevents.PacketEvents;
import com.github.retrooper.packetevents.PacketEventsAPI;
import io.github.fisher2911.hmcleaves.cache.BlockCache;
import io.github.fisher2911.hmcleaves.cache.ChunkBlockCache;
import io.github.fisher2911.hmcleaves.cache.WorldBlockCache;
import io.github.fisher2911.hmcleaves.command.LeavesCommand;
import io.github.fisher2911.hmcleaves.config.LeavesConfig;
import io.github.fisher2911.hmcleaves.data.LeafDatabase;
import io.github.fisher2911.hmcleaves.debug.Debugger;
import io.github.fisher2911.hmcleaves.hook.Hooks;
import io.github.fisher2911.hmcleaves.listener.InteractionListener;
import io.github.fisher2911.hmcleaves.listener.LeafAndLogEditListener;
import io.github.fisher2911.hmcleaves.listener.LeafDropListener;
import io.github.fisher2911.hmcleaves.listener.PlayerJoinListener;
import io.github.fisher2911.hmcleaves.listener.SoundListener;
import io.github.fisher2911.hmcleaves.listener.WorldAndChunkLoadListener;
import io.github.fisher2911.hmcleaves.nms.FeatureHandler;
import io.github.fisher2911.hmcleaves.packet.BlockBreakManager;
import io.github.fisher2911.hmcleaves.packet.LeavesPacketListener;
import io.github.retrooper.packetevents.factory.spigot.SpigotPacketEventsBuilder;
import org.bstats.bukkit.Metrics;
import org.bukkit.Bukkit;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.java.JavaPlugin;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class HMCLeaves extends JavaPlugin {

    private LeavesConfig leavesConfig;
    private BlockCache blockCache;
    private LeafDatabase leafDatabase;
    private BlockBreakManager blockBreakManager;
    private WorldAndChunkLoadListener worldAndChunkLoadListener;
    private LeavesPacketListener leavesPacketListener;
    private FeatureHandler featureHandler;

    @Override
    public void onLoad() {
        this.leavesConfig = new LeavesConfig(
                this,
                new ConcurrentHashMap<>(),
                new ConcurrentHashMap<>(),
                new HashMap<>(),
                new HashMap<>(),
                new HashMap<>(),
                new HashMap<>()
        );
        this.blockCache = new BlockCache(new ConcurrentHashMap<>());
        try {
            final String version = Bukkit.getServer().getVersion();
            if (version.contains("1.18.2")) {
                this.createFeatureHandler("v1_18_2");
            } else if (version.contains("1.18")) {
                this.createFeatureHandler("v1_18");
            }
            if (version.contains("1.19.4")) {
                this.createFeatureHandler("v1_19_4");
            } else if (version.contains("1.19.3")) {
                this.createFeatureHandler("v1_19_3");
            } else if (version.contains("1.19")) {
                this.createFeatureHandler("v1_19");
            }
            if (version.contains("1.20")) {
                this.createFeatureHandler("v1_20");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void createFeatureHandler(String packageVersion) throws ClassNotFoundException, IllegalAccessException, InstantiationException, InvocationTargetException {
        this.featureHandler = (FeatureHandler) Class.forName("io.github.fisher2911." + packageVersion + ".FeatureHandler").getConstructors()[0]
                .newInstance(this.leavesConfig, this.blockCache);
        this.featureHandler.register();
    }

    @Override
    public void onEnable() {
        final PacketEventsAPI<Plugin> api = SpigotPacketEventsBuilder.build(this);
        api.getSettings().checkForUpdates(false);
        PacketEvents.setAPI(api);
        PacketEvents.getAPI().load();
        PacketEvents.getAPI().init();
        this.leafDatabase = new LeafDatabase(this);
        this.blockBreakManager = new BlockBreakManager(new ConcurrentHashMap<>(), this);
        this.worldAndChunkLoadListener = new WorldAndChunkLoadListener(this);
        this.leavesPacketListener = new LeavesPacketListener(this);
        this.registerPacketListeners();
        this.registerListeners();
        Hooks.load(this);
        this.leavesConfig.load();
        this.leafDatabase.load();
        Bukkit.getScheduler().runTaskLater(this, () -> this.worldAndChunkLoadListener.loadDefaultWorlds(), 20);
        this.getCommand("hmcleaves").setExecutor(new LeavesCommand(this));
        final int bStatsPluginId = 16900;
        final Metrics metrics = new Metrics(this, bStatsPluginId);
    }

    public void reload() {
        this.leavesConfig.reload();
    }

    @Override
    public void onDisable() {
        this.leafDatabase.shutdownNow().forEach(Runnable::run);
        for (var entry : this.blockCache.getCache().entrySet()) {
            final WorldBlockCache worldBlockCache = entry.getValue();
            for (var chunkEntry : worldBlockCache.getBlockCacheMap().entrySet()) {
                final ChunkBlockCache chunkBlockCache = chunkEntry.getValue();
                if (chunkBlockCache.isClean()) continue;
                this.leafDatabase.saveBlocksInChunk(chunkBlockCache);
            }
        }
        Debugger.getInstance().shutdown();
    }

    private void registerPacketListeners() {
        PacketEvents.getAPI().getEventManager().registerListener(this.leavesPacketListener);
    }

    private void registerListeners() {
        List.of(
                        new InteractionListener(this),
                        new LeafDropListener(this),
                        this.worldAndChunkLoadListener,
                        new LeafAndLogEditListener(this),
                        new SoundListener(this),
                        new PlayerJoinListener(this)
                )
                .forEach(listener -> getServer().getPluginManager().registerEvents(listener, this));
    }

    public LeavesConfig getLeavesConfig() {
        return this.leavesConfig;
    }

    public BlockCache getBlockCache() {
        return this.blockCache;
    }

    public LeafDatabase getLeafDatabase() {
        return this.leafDatabase;
    }

    public BlockBreakManager getBlockBreakManager() {
        return this.blockBreakManager;
    }

    public LeavesPacketListener getLeavesPacketListener() {
        return this.leavesPacketListener;
    }

    public boolean debug() {
        return true;
    }

}

File: common/src/main/kotlin/com/hibiscusmc/hmcleaves/HMCLeaves.kt
package com.hibiscusmc.hmcleaves

import com.github.retrooper.packetevents.PacketEvents
import com.hibiscusmc.hmcleaves.block.BlockChecker
import com.hibiscusmc.hmcleaves.command.HMCLeavesCommand
import com.hibiscusmc.hmcleaves.config.LeavesConfig
import com.hibiscusmc.hmcleaves.database.LeavesDatabase
import com.hibiscusmc.hmcleaves.debug.ActiveLeavesLogger
import com.hibiscusmc.hmcleaves.debug.DisabledLeavesLogger
import com.hibiscusmc.hmcleaves.debug.LeavesLogger
import com.hibiscusmc.hmcleaves.hook.Hooks
import com.hibiscusmc.hmcleaves.listener.BukkitListeners
import com.hibiscusmc.hmcleaves.listener.ChunkListener
import com.hibiscusmc.hmcleaves.listener.SoundListener
import com.hibiscusmc.hmcleaves.nms.NMSHandler
import com.hibiscusmc.hmcleaves.packet.PacketListener
import com.hibiscusmc.hmcleaves.packet.mining.BlockBreakManager
import com.hibiscusmc.hmcleaves.user.UserManager
import com.hibiscusmc.hmcleaves.world.WorldManager
import io.github.retrooper.packetevents.factory.spigot.SpigotPacketEventsBuilder
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import java.lang.reflect.InvocationTargetException


class HMCLeaves : JavaPlugin() {

    val leavesConfig: LeavesConfig by lazy { LeavesConfig(this) }
    val worldManager = WorldManager()
    private lateinit var database: LeavesDatabase
    val blockBreakManager: BlockBreakManager by lazy { BlockBreakManager(plugin = this) }
    val userManager: UserManager by lazy { UserManager(this) }
    private lateinit var leavesLogger: LeavesLogger
    private val blockChecker = BlockChecker(this, worldManager)
    private lateinit var nmsHandler: NMSHandler

    override fun onLoad() {
        PacketEvents.setAPI(SpigotPacketEventsBuilder.build(this))
        PacketEvents.getAPI().settings.reEncodeByDefault(true)
            .checkForUpdates(false)
            .debug(true)
        PacketEvents.getAPI().load()
        if (Bukkit.getVersion().contains("1.19")) {
            this.nmsHandler = this.createNMSHandler("v1_19")
        }
        if (Bukkit.getVersion().contains("1.20.4")) {
            this.nmsHandler = this.createNMSHandler("v1_20_4")
        }
        if (Bukkit.getVersion().contains("1.21")) {
            this.nmsHandler = this.createNMSHandler("v1_21")
        }
    }

    @Throws(
        ClassNotFoundException::class,
        IllegalAccessException::class,
        InstantiationException::class,
        InvocationTargetException::class
    )
    private fun createNMSHandler(packageVersion: String): NMSHandler {
        return Class.forName("com.hibiscusmc.hmcleaves.${packageVersion}.NMSHandler").constructors[0]
            .newInstance() as NMSHandler
    }

    override fun onEnable() {
        this.leavesConfig.load()
        this.leavesLogger = if (this.leavesConfig.sendDebugMessages()) {
            ActiveLeavesLogger(this)
        } else {
            DisabledLeavesLogger
        }
        this.leavesLogger.init()
        this.database = LeavesDatabase.createDatabase(this, this.leavesConfig.getDatabaseSettings().type)

        val pluginManager = this.server.pluginManager
        pluginManager.registerEvents(BukkitListeners(this), this)
        pluginManager.registerEvents(ChunkListener(this), this)
        pluginManager.registerEvents(SoundListener(this), this)

        PacketEvents.getAPI().eventManager.registerListener(PacketListener(this, this.blockChecker))
        PacketEvents.getAPI().init()

        this.registerCommands()

        this.database.init()

        this.blockChecker.start()

        // load hooks after all plugins have been loaded, cannot use
        // softdepend because HMCLeaves must load before the world loads
        Bukkit.getScheduler().runTaskLater(this, Runnable {
            Hooks.init()
            // load textures after hooks have been initialized
            Hooks.reloadTextures()
            this.leavesConfig.loadTextures()
        }, 1)
    }

    private fun registerCommands() {
        this.getCommand("hmcleaves")?.setExecutor(HMCLeavesCommand(this))
    }

    override fun onDisable() {
        this.blockChecker.stop()
        this.leavesLogger.sendSynced(false)
        PacketEvents.getAPI().terminate()
        for (world in Bukkit.getWorlds()) {
            this.database.saveWorld(world, true)
        }
    }

    fun reload() {
        this.leavesConfig.reload()
        this.worldManager.reload(this.leavesConfig)
    }

    fun getDatabase(): LeavesDatabase = this.database

    fun getLeavesLogger() = this.leavesLogger

    fun getBlockChecker() = this.blockChecker

    fun getNMSHandler() = this.nmsHandler

}